bundle-workflow/python/assemble_workflow/bundle_recorder.py

```diff
@@ -0,0 +1,82 @@
+
+import os
+import yaml
+from manifests.bundle_manifest import BundleManifest
+from urllib.parse import urljoin
+
+
+class BundleRecorder:
+    def __init__(self, build, output_path, artifacts_dir, public_url=None):
+        self.build_id = build.id
+        self.public_url = public_url
+        self.version = build.version
+        self.output_path = output_path
+        self.tar_name = os.path.basename(output_path)
+        self.artifacts_dir = artifacts_dir
+        self.bundle_manifest = self.BundleManifestBuilder(build.id, build.name, build.version, build.architecture, self.get_tar_location())
+
+    def get_public_url_path(self, folder, rel_path):
+        path = "{}/{}/{}/{}".format(folder, self.version, self.build_id, rel_path)
+        return urljoin(self.public_url, path)
+
+    def get_location(self, folder_name, rel_path, abs_path):
+        if self.public_url:
+            return self.get_public_url_path(folder_name, rel_path)
+        return abs_path
+
+    # Assembled bundles are expected to be served from a separate "bundles" folder
+    # Example: https://artifacts.opensearch.org/bundles/1.0.0/<build-id
+    def get_tar_location(self):
+        return self.get_location("bundles", self.tar_name, self.output_path)
+
+    # Build artifacts are expected to be served from a "builds" folder
+    # Example: https://artifacts.opensearch.org/builds/1.0.0/<build-id>
+    def get_component_location(self, component_rel_path):
+        abs_path = os.path.join(self.artifacts_dir, component_rel_path)
+        return self.get_location("builds", component_rel_path, abs_path)
+
+    def record_component(self, component, rel_path):
+        abs_path = os.path.join(self.artifacts_dir, rel_path)
+        self.bundle_manifest.append_component(component.name, component.repository, component.ref,
+                                              component.commit_id, self.get_component_location(rel_path))
+
+    def get_manifest(self):
+        return self.bundle_manifest.to_manifest()
+
+    def write_manifest(self, folder):
+        output_manifest = self.get_manifest()
+        manifest_path = os.path.join(folder, 'manifest.yml')
+        with open(manifest_path, 'w') as file:
+            yaml.dump(output_manifest.to_dict(), file)
+
+    class BundleManifestBuilder:
+        def __init__(self, build_id, name, version, arch, location):
+            self.data = {}
+            self.data['build'] = {}
+            self.data['build']['id'] = build_id
+            self.data['build']['name'] = name
+            self.data['build']['version'] = str(version)
+            self.data['build']['architecture'] = arch
+            self.data['build']['location'] = location
+            self.data['schema-version'] = '1.0'
+            # We need to store components as a hash so that we can append artifacts by component name
+            # When we convert to a BundleManifest this will get converted back into a list
+            self.data['components_hash'] = {}
```

Comment:
Do you need to store the components as a hash? The bundler does not need to update components, so can you store them directly as a list and avoid the munging on lines 80-81?